cmd/embedded-cluster/reset.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return err
+	}
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(adminConfig)
+	if err != nil {
+		return err
+	}
+	h.Kclient, err = client.New(restConfig, client.Options{})
+	autopilot.AddToScheme(h.Kclient.Scheme())
+	if err != nil {
+		return fmt.Errorf("couldn't create k8s config: %w", err)
+	}
+	return nil
+}
+
+// getHostName fetches the hostname for the node
+func (h *hostInfo) getHostName() error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil
+	}
+	h.Hostname = hostname
+	return nil
+}
+
+// isControlPlane attempts to determine if the node is a controlplane node
+func (h *hostInfo) isControlPlane() bool {
+	labels := h.Node.GetLabels()
+	return labels["node-role.kubernetes.io/control-plane"] == "true"
+}
+
+// getNodeObject fetches the node object from the k8s api server
+func (h *hostInfo) getNodeObject(ctx context.Context) error {
+	err := h.Kclient.Get(ctx, client.ObjectKey{Name: h.Hostname}, &h.Node)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getControlNodeObject fetches the controlNode object from the k8s api server
+func (h *hostInfo) getControlNodeObject(ctx context.Context) error {
+	err := h.Kclient.Get(ctx, client.ObjectKey{Name: h.Hostname}, &h.ControlNode)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *hostInfo) checkQuorumSafety(c *cli.Context) (bool, string, error) {
+	if c.Bool("yes-really-reset") {
+		return true, "", nil
+	}
+	out, err := exec.Command(k0s, "etcd", "member-list").Output()
+	if err != nil {
+		return false, "", fmt.Errorf("unable to fetch etcd member list, %w, %s", err, out)
+	}
+	etcd := etcdMembers{}
+	err = json.Unmarshal(out, &etcd)
+	if err != nil {
+		return false, "", fmt.Errorf("unable to unmarshal etcd member list, %w, %s", err, out)
+	}
+	if len(etcd.Members) < 3 {
+		return true, "", nil
+	}
+	if len(etcd.Members) == 3 {
+		return false, "cluster has 3 control-plane nodes, removing this node will cause etcd to lose quorum", nil
+	}
+	if len(etcd.Members)%2 != 0 {
+		return false, "cluster would have even number of control-plane nodes after resetting this node, this could cause etcd to become unstable", nil
+	}
+	return true, "", nil
+}
+
+// leaveEtcdcluster uses k0s to attempt to leave the etcd cluster
+func (h *hostInfo) leaveEtcdcluster() error {
+	out, err := exec.Command(k0s, "etcd", "leave").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("unable to leave etcd cluster: %w, %s", err, string(out))
+	}
+	return nil
+}
+
+// stopK0s attempts to stop the k0s service
+func (h *hostInfo) stopAndResetK0s() error {
+	out, err := exec.Command(k0s, "stop").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("could not stop k0s service: %w, %s", err, string(out))
+	}
+	out, err = exec.Command(k0s, "reset").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("could not reset k0s: %w, %s", err, string(out))
+	}
+	fmt.Println("Node has been reset, please reboot to ensure transient configuration is also reset")
```

Comment:
Node has been reset. Please reboot to ensure transient configuration is also reset.